src/Sulu/Component/Persistence/EventSubscriber/ORM/MetadataSubscriber.php

```diff
@@ -0,0 +1,127 @@
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Persistence\EventSubscriber\ORM;
+
+use Doctrine\Common\EventSubscriber;
+use Doctrine\ORM\Configuration;
+use Doctrine\ORM\Event\LoadClassMetadataEventArgs;
+use Doctrine\ORM\Events;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\Mapping\ClassMetadataInfo;
+
+/**
+ * Doctrine subscriber used to manipulate metadata.
+ */
+class MetadataSubscriber implements EventSubscriber
+{
+    /**
+     * @var array
+     */
+    protected $classes;
+
+    /**
+     * Constructor
+     *
+     * @param array $classes
+     */
+    public function __construct($classes)
+    {
+        $this->classes = $classes;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getSubscribedEvents()
+    {
+        return array(
+            Events::loadClassMetadata,
+        );
+    }
+
+    /**
+     * Load the class data, mapping the created and changed fields
+     * to datetime fields.
+     *
+     * @param LoadClassMetadataEventArgs $event
+     */
+    public function loadClassMetadata(LoadClassMetadataEventArgs $event)
+    {
+        /** @var ClassMetadataInfo $metadata */
+        $metadata = $event->getClassMetadata();
+
+        $this->process($metadata);
+
+        if (!$metadata->isMappedSuperclass) {
+            $this->setAssociationMappings($metadata, $event->getEntityManager()->getConfiguration());
+        } else {
+            $this->unsetAssociationMappings($metadata);
+        }
+    }
+
+    private function process(ClassMetadataInfo $metadata)
+    {
+        foreach ($this->classes as $application => $classes) {
+            foreach ($classes as $class) {
+                if (isset($class['model']) && $class['model'] === $metadata->getName()) {
+                    $metadata->isMappedSuperclass = false;
+
+                    if (isset($class['repository'])) {
+                        $metadata->setCustomRepositoryClass($class['repository']);
+                    }
+                }
+            }
+        }
+    }
+
+    private function setAssociationMappings(ClassMetadataInfo $metadata, Configuration $configuration)
+    {
+        foreach (class_parents($metadata->getName()) as $parent) {
+            $parentMetadata = new ClassMetadata(
+                $parent,
+                $configuration->getNamingStrategy()
+            );
+
+            if (in_array($parent, $configuration->getMetadataDriverImpl()->getAllClassNames())) {
+                $configuration->getMetadataDriverImpl()->loadMetadataForClass($parent, $parentMetadata);
+                if ($parentMetadata->isMappedSuperclass) {
+                    foreach ($parentMetadata->getAssociationMappings() as $key => $value) {
+                        if ($this->hasRelation($value['type'])) {
+                            $metadata->associationMappings[$key] = $value;
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
This can be de-nested:

if (!in_array(..) {
    return
}

if (!$parentMetadata->isMappedSuprtclass() {
    return;
}

foreach ( ... ) {
}
````